pi5/pinctrl_utils.go

```diff
@@ -0,0 +1,287 @@
+func cleanFilePath(childNodePath string) string {
+	childNodePath = strings.TrimSpace(childNodePath)
+	re := regexp.MustCompile(`[\x00-\x1F\x7F-\x9F]`) // gets rid of random non printable chars. works for now but make cleaner later
+	childNodePath = re.ReplaceAllString(childNodePath, "")
+	return childNodePath
+}
+
+// Reads the /reg file and converts the bytestream into a uint64 representing the GPIO Node's physical address within its parent's space. (pre address mapping)
+func getRegAddr(childNodePath string, numPAddrCells uint32) (uint64, error) {
+
+	childNodePath += "/reg"
+	childNodePath = cleanFilePath(childNodePath)
+
+	regByteContents, err := os.ReadFile(childNodePath)
+	if err != nil {
+		return INVALID_ADDR, fmt.Errorf("trouble getting reg addr info for %s: %w\n", childNodePath, err)
+	}
+
+	physAddr := INVALID_ADDR
+	switch numPAddrCells {
+	case 1: // reading in 32 bits. regardless we must convert to a 64 bit address so we add a bunch of 0s to the beginning.
+		physAddr = uint64(binary.BigEndian.Uint32(regByteContents[:(4 * numPAddrCells)]))
+	case 2: // reading in 64 bits
+		physAddr = binary.BigEndian.Uint64(regByteContents[:(4 * numPAddrCells)])
+	case 3: // reading in more than 64 bits. we only want the last 64 bits of the address though, so we cut off the other portion of it
+		physAddr = binary.BigEndian.Uint64(regByteContents[(4 * (numPAddrCells - 2)):(4 * numPAddrCells)])
+	}
+	return physAddr, err
+}
+
+// Reads the /ranges file and converts the bytestream into integers representing the < child address parent address parent size >
+func getRangesAddrInfo(childNodePath string, numCAddrCells uint32, numPAddrCells uint32, numAddrSpaceCells uint32) ([]rangeInfo, error) {
+
+	childNodePath += "/ranges"
+	childNodePath = cleanFilePath(childNodePath)
+	var addrRanges []rangeInfo
+
+	rangeByteContents, err := os.ReadFile(childNodePath)
+	if err != nil {
+		return addrRanges, fmt.Errorf("trouble getting reg addr info for %s: %w\n", childNodePath, err)
+	}
+
+	numRanges := uint32(len(rangeByteContents)) / (4 * (numCAddrCells + numPAddrCells + numAddrSpaceCells))
+
+	for i := uint32(0); i < numRanges; i++ {
+
+		childAddr, parentAddr := INVALID_ADDR, INVALID_ADDR
+		addrSpaceSize := uint64(0)
+
+		switch numCAddrCells {
+		case 1: // reading in 32 bits. regardless we must convert to a 64 bit address so we add a bunch of 0s to the beginning.
+			childAddr = uint64(binary.BigEndian.Uint32(rangeByteContents[:(4 * numCAddrCells)]))
+		case 2: // reading in 64 bits
+			childAddr = binary.BigEndian.Uint64(rangeByteContents[:(4 * numCAddrCells)])
+		case 3: // reading in more than 64 bits. we only want the last 64 bits of the address though, so we cut off the other portion of it
+			childAddr = binary.BigEndian.Uint64(rangeByteContents[(4 * (numCAddrCells - 2)):(4 * numCAddrCells)])
+		}
+		rangeByteContents = rangeByteContents[(4 * numCAddrCells):] // flush the bytes already parsed out of the array
+
+		switch numPAddrCells {
```

Comment:
shouldve said this in other PR, can you make a helper function for these switch statements since theyre all the same?